keps/sig-apps/0034-maxunavailable-for-statefulsets.md

```diff
@@ -0,0 +1,179 @@
+---
+kep-number: 34
+title: Implement maxUnavailable for StatefulSets
+authors:
+  - "@krmayankk"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-apps
+reviewers:
+  - "@janetkuo"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2018-12-29
+last-updated: 2018-12-29
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# Implement maxUnavailable in StatefulSet
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories [optional]](#user-stories-optional)
+      * [Story 1](#story-1)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+The purpose of this enhancement is to implement maxUnavailable for StatefulSet during RollingUpdate. When a StatefulSet’s 
+`.spec.updateStrategy.type` is set to `RollingUpdate`, the StatefulSet controller will delete and recreate each Pod 
+in the StatefulSet. The updating of each Pod currently happens one at a time. With support for `maxUnavailable`, the updating
+will proceed `maxUnavailable` number of pods at a time.
+
+
+## Motivation
+
+Consider the following scenarios:-
+
+1: My containers publish metrics to a time series system. If I am using a Deployment, each rolling update creates a new pod name and hence the metrics 
+published by these new pod starts a new time series which makes tracking metrics for the application difficult. While this could be mitigated, 
+it requires some tricks on the time series collection side. It would be so much better, If we could use a StatefulSet object so my object names doesnt 
```

Comment:
Is this fixed if Prometheus is able to treat all pods under a service as a single source of metrics? Sounds like the right way to fix this issue.

Reply:
we want Prometheus to treat a service as composed of `n` distinct entities sending metrics(where `n` is the number of replicas in the service) and not as `n *(number of rolling updates to the service)`  distinct entities sending metrics to Prometheus.  Imagine that node-0 is sending metrics and after upgrade node-0 starts sending metrics with some weird behavior which allows you to know node-0 started misbehaving after upgrade. With deployments node-xaqr was sending metrics and now node-uehd is sending metrics.